tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_canonicalize_reduction.cc

```diff
@@ -0,0 +1,238 @@
+        // As a workaround, we convert the `reduce to scalar` to a rank 2
+        // column reduction having following form:
+        // Suppose nelems = ProdutionOp(ShapeOp(I)), We convert I into
+        // shape `[nelems, 1]`.
+        // TODO(disc): this may have performance issue. Implements a reduce to
+        // scalar schedule if necessnary.
+        new_operand_dims.push_back(nelem_to_reduce);
+        new_operand_dims.push_back(nelem_to_keep);
+        attr = DenseIntElementsAttr::get(
+            RankedTensorType::get({1}, b.getIntegerType(64)), {0ll});
+      } else if (dims_to_reduce[0] == 0) {
+        // case a) column reduction
+        new_operand_dims.push_back(nelem_to_reduce);
+        new_operand_dims.push_back(nelem_to_keep);
+        attr = DenseIntElementsAttr::get(
+            RankedTensorType::get({1}, b.getIntegerType(64)), {0ll});
+      } else {
+        // case b) raw reduction
+        new_operand_dims.push_back(nelem_to_keep);
+        new_operand_dims.push_back(nelem_to_reduce);
+        attr = DenseIntElementsAttr::get(
+            RankedTensorType::get({1}, b.getIntegerType(64)), {1ll});
+      }
+
+      Value new_operand_shape =
+          b.create<tensor::FromElementsOp>(loc, new_operand_dims);
+
+      SmallVector<Value, 4> new_operands;
+      for (Value operand : op.inputs()) {
+        new_operands.push_back(b.create<DynamicReshapeOp>(
+            loc,
+            RankedTensorType::get(
+                SmallVector<int64_t, 4>(new_operand_dims.size(),
+                                        ShapedType::kDynamicSize),
+                elem_ty),
+            operand, new_operand_shape));
+      }
+      auto new_op =
+          b.create<ReduceOp>(loc, new_operands, op.init_values(), attr);
+      BlockAndValueMapping mapping;
+      op.body().cloneInto(&new_op.body(), new_op.body().end(), mapping);
+
+      SmallVector<Value, 4> new_results;
+      if (dims_to_keep.empty()) {
+        // case c) reduce to scalar
+        // reshape rank 1 tensor with size 1 to a rank 0 tensor
+        for (Value result : new_op.getResults()) {
+          new_results.push_back(b.create<ReshapeOp>(
+              loc, RankedTensorType::get({}, elem_ty), result));
+        }
+      } else {
+        SmallVector<Value, 4> result_dims;
+        for (int64_t i : dims_to_keep) {
+          Value dim_index = b.create<memref::DimOp>(loc, op.getOperand(0), i);
+          result_dims.push_back(
+              b.create<IndexCastOp>(loc, dim_index, shape_scalar_type));
+        }
+        Value result_shape = b.create<tensor::FromElementsOp>(loc, result_dims);
+        for (auto&& e : llvm::zip(op.getResults(), new_op.getResults())) {
+          new_results.push_back(b.create<DynamicReshapeOp>(
+              loc, std::get<0>(e).getType(), std::get<1>(e), result_shape));
+        }
+      }
+
+      for (auto&& e : llvm::zip(op.getResults(), new_results)) {
+        std::get<0>(e).replaceAllUsesWith(std::get<1>(e));
+      }
```

Comment:
```suggestion
      op.replaceAllUsesWith(new_op);
```

Reply:
I think can't directly use `op.replaceAllUsesWith()` here, because we create a Reshape / DynamicReshape Op behind the new_op.

Reply:
Of course...